kevm-pyk/src/kevm_pyk/kproj/evm-semantics/optimizations.md

```diff
@@ -315,6 +315,138 @@ module EVM-OPTIMIZATIONS
      andBool ( size( WS ) <=Int 1023 )
      [priority(40)]
 
+  rule
+    <kevm>
+      <k>
+        ( #next [ ISZERO ] => .K ) ...
+      </k>
+      <schedule>
+        SCHED
+      </schedule>
+      <ethereum>
+        <evm>
+          <callState>
+            <wordStack>
+              ( ListItem(W0) WS => pushList(bool2Word( W0 ==Int 0 ), WS) )
+            </wordStack>
+            <pc>
+              ( PCOUNT => ( PCOUNT +Int 1 ) )
+            </pc>
+            <gas>
+              ( GAVAIL => ( GAVAIL -Gas Gverylow < SCHED > ) )
+            </gas>
+            ...
+          </callState>
+          ...
+        </evm>
+        ...
+      </ethereum>
+      ...
+    </kevm>
+    requires ( Gverylow < SCHED > <=Gas GAVAIL )
+     andBool ( size( WS ) <=Int 1023 )
+     [priority(40)]
+
+  rule
+    <kevm>
+      <k>
+        ( #next [ JUMPDEST ] => .K ) ...
+      </k>
+      <schedule>
+        SCHED
+      </schedule>
+      <ethereum>
+        <evm>
+          <callState>
+            <pc>
+              ( PCOUNT => ( PCOUNT +Int 1 ) )
+            </pc>
+            <gas>
+              ( GAVAIL => ( GAVAIL -Gas Gjumpdest < SCHED > ) )
+            </gas>
+            ...
+          </callState>
+          ...
+        </evm>
+        ...
+      </ethereum>
+      ...
+    </kevm>
+    requires ( Gjumpdest < SCHED > <=Gas GAVAIL )
+     [priority(40)]
+
+  rule
+    <kevm>
+      <k>
+        ( #next [ JUMP ] => .K ) ...
+      </k>
+      <schedule>
+        SCHED
+      </schedule>
+      <ethereum>
+        <evm>
+          <callState>
+            <jumpDests>
+              DESTS
+            </jumpDests>
+            <wordStack>
+              ( ListItem(W0) WS => WS )
+            </wordStack>
+            <pc>
+              ( PCOUNT => W0 )
+            </pc>
+            <gas>
+              ( GAVAIL => ( GAVAIL -Gas Gmid < SCHED > ) )
+            </gas>
+            ...
+          </callState>
+          ...
+        </evm>
+        ...
+      </ethereum>
+      ...
+    </kevm>
+    requires ( Gmid < SCHED > <=Gas GAVAIL )
+     andBool ( size( WS ) <=Int 1023 )
+     andBool W0 in DESTS
+     [priority(40)]
+
+  rule
+    <kevm>
+      <k>
+        ( #next [ JUMPI ] => .K ) ...
+      </k>
+      <schedule>
+        SCHED
+      </schedule>
+      <ethereum>
+        <evm>
+          <callState>
+            <jumpDests>
+              DESTS
+            </jumpDests>
+            <wordStack>
+              ( ListItem(W0) ListItem(W1) WS => WS )
+            </wordStack>
+            <pc>
+              ( PCOUNT => #if W1 =/=Int 0 #then W0 #else PCOUNT +Int 1 #fi )
+            </pc>
+            <gas>
+              ( GAVAIL => ( GAVAIL -Gas Ghigh < SCHED > ) )
+            </gas>
+            ...
+          </callState>
+          ...
+        </evm>
+        ...
+      </ethereum>
+      ...
+    </kevm>
+    requires ( Ghigh < SCHED > <=Gas GAVAIL )
+     andBool ( size( WS ) <=Int 1023 )
```

Comment:
If I understand correctly, there is a mistake in this requires.

```suggestion
     andBool ( size( WS ) <=Int 1022 )
```

Reply:
So, you might be right, but I'm actually pretty sure that these overflow checks are only required on opcodes that emit more stack values than they consume, of which this one is not one. Let me keep tinkering and I may end up making a modified version of this PR that just deletes the unnecessary checks.